### ISSUE TITLE
Derive subscription startup timeout from the number of models

### ISSUE THREAD

**Issue by rjuliano** [aws-amplify/amplify-android#854]
Currently, subscription startup timeout is set to 10 seconds. We need to change this so we can take into account the number of models in the model provider

**Comment by jamesonwilliams** [comment]
Carrying over [some discussion from Discord, 01SEP2020](https://discordapp.com/channels/705853757799399426/735987596462653461/750261715287146606):

@saltonmassally:
> For projects with models > 10, timing out at 10 second is problematic and subscriptions startup consistently times out. Can we increase the timeout threshold?

@jamesonwilliams:
> The current code establishes the connections serially, not in parallel, right? Yes I think we could make it dynamic (k * modelCount), or even just user-configurable?

@saltonmassally:
> yes it is serial, i thought about parallelising it as i see no reason why we can't do that.. also making it dynamic might be a good short-term fix while we throw user-configurability in the backlog
Let me know your thoughts, I can open a PR accordingly

@jamesonwilliams 
> I think I originally had it in parallel; I believe I made it serial only to simplify the error reporting from Rx. Otherwise, if multiple things fail in parallel, you have to sort out which ones to report -- or wait for all the errors and report multiple errors, etc. However, if we can get that implemented cleanly, it would be a great optimization.
> I think either of those solutions could be an improvement. User-configurable might be more explicit, and less confusing? @richardmcclellan @rjuliano What do you guys think of the options here?

@rjuliano 
> I think it's OK in the short-term to derive that value based on the number of models. Adding a configurable setting would be good long term, but it probably needs to be uniform across all platforms.

@richardmcclellan 
> Maybe I'm not understanding the code. Isn't it already in parallel, because we use `Observable.merge` [here](https://github.com/aws-amplify/amplify-android/blob/4d426937c2e0a66345f7c63ca02d5c6f3f675cc5/aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/SubscriptionProcessor.java#L103).   Serial would be `Observable.concat` I think, right?

@rjuliano
> Regarding parallelizing it, I believe we currently do that. The list of observables is constructed serially, however, the [Observable.merge](https://github.com/aws-amplify/amplify-android/blob/main/aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/SubscriptionProcessor.java#L103) call is set on run on the IO scheduler

@saltonmassally 
> So for now let's just derive that value based on the number of models. any recommendation of an equation for that? or rather a formula
another consideration how much does connectivity affect how long it takes for a subscription to be made?

@jamesonwilliams 
> Hm, if we are already opening the subscriptions in parallel, then the timeout only need be greater than the longest single setup, right? (This is like Amdahl's Law.) So, doing modelCount * constant would be too aggressive?

@rjuliano
> Sort of...we may still have some sort of bottleneck going on with the websocket client connection. It's one connection shared for all subscriptions I believe. At least that was the recommendation from the AppSync docs.

**Comment by jamesonwilliams** [comment]
Resolved by #858, closing.

**Comment by fjnoyp** [comment]
Hi during our testing of Flutter Datastore on Android, we noticed that the app would not sync all models when our stored dataset was large (3000+ stored models).  

We used the following ModelSchema.graphql: 

`type Blog @model {
  id: ID!
  name: String!
  posts: [Post] @connection(keyName: "byBlog", fields: ["id"])
}

type Post @model @key(name: "byBlog", fields: ["blogID"]) {
  id: ID!
  title: String!
  rating: Int
  created: AWSDateTime

  blogID: ID!
  blog: Blog @connection(fields: ["blogID"])
  comments: [Comment] @connection(keyName: "byPost", fields: ["id"])
}

type Comment @model @key(name: "byPost", fields: ["postID", "content"]) {
  id: ID!
  postID: ID!
  post: Post @connection(fields: ["postID"])
  content: String!
}



type HasOneModel @model{
  id: ID!
  title: String!
  child: BelongsToModel @connection
}

type BelongsToModel @model{
  id: ID!
  title: String! 
}

enum TestEnum { 
  yes,
  no,
  maybe
}

type AllTypeModel @model { 
  id: ID!
  stringType: String!
  intType: Int!
  floatType: Float!
  boolType: Boolean!
  dateTimeType: AWSDateTime!
  timestampType: AWSTimestamp!
  enumType: TestEnum 
}`

However we only used: 
Post, Comment, Blog, AllTypeModel 

Here are the logs.  Please note there is another error here - the app crashes when it loses internet connection, I believe that's in another issue being tracked already.

[AndroidLogs.txt](https://github.com/aws-amplify/amplify-android/files/5640420/AndroidLogs.txt)

Please let me know if there is more information we can provide

**Comment by richardmcclellan** [comment]
> Hi during our testing of Flutter Datastore on Android, we noticed that the app would not sync all models when our stored dataset was large (3000+ stored models).

This should be resolved by https://github.com/aws-amplify/amplify-android/pull/1161 which raised the Orchestrator timeout from 10 seconds to 60 seconds. 

> Please note there is another error here - the app crashes when it loses internet connection, I believe that's in another issue being tracked already.

Yes, this issue was resolved by https://github.com/aws-amplify/amplify-android/pull/1039
